File: src/main/java/com/service/serviceImpl/StatisticsComputationImpl.java
package com.service.serviceImpl;

import com.model.Statistics;
import com.service.StatisticsComputation;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.DoubleSummaryStatistics;
import java.util.List;
import java.util.stream.Collectors;

@Service
public class StatisticsComputationImpl implements StatisticsComputation {

    public Statistics computeStatistics(List<BigDecimal> amounts){
        DoubleSummaryStatistics summaryStatistics =
                amounts.stream()
                       .collect(Collectors.summarizingDouble(BigDecimal::doubleValue));

        return mapStatistics(summaryStatistics);
    }

    private Statistics mapStatistics(DoubleSummaryStatistics summaryStatistics) {
        BigDecimal sum = toScaledRoundedBigDecimal(summaryStatistics.getSum());
        BigDecimal avg = toScaledRoundedBigDecimal(summaryStatistics.getAverage());
        BigDecimal max = getFiniteValue(summaryStatistics.getMax());
        BigDecimal min = getFiniteValue(summaryStatistics.getMin());
        long count = summaryStatistics.getCount();

        return new Statistics(sum, avg, max, min, count);
    }

    private BigDecimal getFiniteValue(double value) {
        return Double.isFinite(value) ? toScaledRoundedBigDecimal(value) : BigDecimal.ZERO.setScale(2);
    }

    private BigDecimal toScaledRoundedBigDecimal(Double value){
        return BigDecimal.valueOf(value).setScale(2, RoundingMode.HALF_UP);
    }

}

File: src/main/kotlin/com/service/serviceImpl/StatisticsComputationImpl.kt
package com.service.serviceImpl

import com.model.Statistics
import com.service.StatisticsComputation
import org.springframework.stereotype.Service
import java.math.BigDecimal
import java.math.RoundingMode
import java.util.*
import java.util.stream.Collectors
import mu.KotlinLogging

private val logger = KotlinLogging.logger {}
@Service
class StatisticsComputationImpl : StatisticsComputation {
    override fun computeStatistics(amounts: List<BigDecimal>): Statistics {
        val summaryStatistics = amounts.stream()
            .collect(Collectors.summarizingDouble( BigDecimal::toDouble ))
        return mapStatistics(summaryStatistics)
    }

    private fun mapStatistics(summaryStatistics: DoubleSummaryStatistics): Statistics {
        logger.debug { "mapStatistics parameters: summaryStatistics: $summaryStatistics" }

        val sum = toScaledRoundedBigDecimal(summaryStatistics.sum)
        val avg = toScaledRoundedBigDecimal(summaryStatistics.average)
        val max = getFiniteValue(summaryStatistics.max)
        val min = getFiniteValue(summaryStatistics.min)
        val count = summaryStatistics.count

        logger.debug { "mapStatistics return: sum: $sum, avg: $avg, max: $max, min: $min, count: $count" }

        return Statistics(sum, avg, max, min, count)
    }

    private fun getFiniteValue(value: Double): BigDecimal {
        return if (value.isFinite()) toScaledRoundedBigDecimal(value) else BigDecimal.ZERO.setScale(2)
    }

    private fun toScaledRoundedBigDecimal(value: Double): BigDecimal {
        return BigDecimal.valueOf(value).setScale(2, RoundingMode.HALF_UP)
    }
}